tools/docker/Dockerfile

```diff
@@ -1,189 +1,112 @@
-RUN if [ "${VIRTUAL_ENV}" = "conda" ] ; then curl ${ANACONDA} -o anaconda.sh && \
-    /bin/bash anaconda.sh -b -p conda && \
-    rm anaconda.sh && \
-    echo ". ${HOME}/conda/etc/profile.d/conda.sh" >> ~/.bashrc && \
-    echo "conda activate base" >> ~/.bashrc ; fi
-
-ENV PATH="${HOME}/${VIRTUAL_ENV}/bin:${PATH}"
-
-# --login option used to source bashrc (thus activating conda env) at every RUN statement
-SHELL ["/bin/bash", "--login", "-c"]
-
-# Python version supported by recommenders
-RUN if [ "${VIRTUAL_ENV}" = "conda" ] ; then conda install python=3.7; fi
-RUN if [ "${VIRTUAL_ENV}" = "venv" ] ; then apt-get -y install python3.7; \
-    apt-get -y install python3-pip; \
-    apt-get -y install python3.7-venv; fi
-RUN if [ "${VIRTUAL_ENV}" = "virtualenv" ] ; then apt-get -y install python3.7; \
-    apt-get -y install python3-pip; \
-    python3.7 -m pip install --user virtualenv; fi
-
-# Activate the virtual environment
-RUN if [ "${VIRTUAL_ENV}" = "venv" ] ; then python3.7 -m venv $HOME/${VIRTUAL_ENV}; \
-    source $HOME/${VIRTUAL_ENV}/bin/activate; \
-    pip install --upgrade pip; \
-    pip install --upgrade setuptools wheel; fi
-RUN if [ "${VIRTUAL_ENV}" = "virtualenv" ] ; then python3.7 -m virtualenv $HOME/${VIRTUAL_ENV}; \
-    source $HOME/${VIRTUAL_ENV}/bin/activate; \
-    pip install --upgrade pip; \
-    pip install --upgrade setuptools wheel; fi
-
-###########
-# CPU Stage
-###########
-FROM base AS cpu
-
-RUN if [ "${VIRTUAL_ENV}" = "venv" ] || [ "${VIRTUAL_ENV}" = "virtualenv" ]; then source $HOME/${VIRTUAL_ENV}/bin/activate; \
-    pip install recommenders[examples]; fi
-RUN if [ "${VIRTUAL_ENV}" = "conda" ] ; then pip install recommenders[examples]; fi
-
-
-###############
-# PySpark Stage
-###############
-FROM base AS pyspark
-
-# Install Java version 8
-RUN apt-get update && \
-    apt-get install -y libgomp1 openjdk-8-jre
+#####################################################################
+# Stage build order depending on the compute, extras, python version
+# and virtual environment:
+#        Compute Stage (CPU/GPU)
+#     -> Base Stage
+#     -> Virtual Environment Stage (Conda/Venv/Virtualenv)
+#     -> Dependencies Stage
+#     -> Final Stage
+#####################################################################
+# Valid computes: cpu, gpu
+ARG COMPUTE="cpu"
+
+
+#####################################################################
+# Compute Stage - CPU
+# Choose an appropriate CPU compute image
+#####################################################################
+# * [buildpack-deps:24.04](https://github.com/docker-library/buildpack-deps/blob/master/ubuntu/noble/Dockerfile)
+#   + [Created on 2024-08-17](https://hub.docker.com/layers/library/buildpack-deps/noble/images/sha256-dbfee7e7ee2340b0d6567efd3a8a9281ce45ee78598485b4d7a7f09fe641811a)
+FROM buildpack-deps@sha256:dbfee7e7ee2340b0d6567efd3a8a9281ce45ee78598485b4d7a7f09fe641811a AS cpu
```

Comment:
what is this? would it be possible to explain what it is and why we are using a specific sha?

Reply:
This base image is built on top of ubuntu by adding the essential tools for building packages such as gcc.  Specifying the sha is for speeding up the image build, otherwise, it will pull the base image every time when there is a new release of the image.

Reply:
got it, nice!